### PR TITLE
Emit events on cm/secret udpates

### PR DESCRIPTION
**Pull request by tomahawk28**
Toward https://github.com/giantswarm/giantswarm/issues/16981

## Checklist

- [x] Update changelog in CHANGELOG.md.

### CODE REVIEW

File: service/watcher/appvalue/configmap.go [hunk 1]
```diff
@@ -104,13 +105,21 @@ func (c *AppValueWatcher) watchConfigMap(ctx context.Context) {
...
+
+				err = c.addAnnotation(ctx, currentApp, cm.GetResourceVersion(), configMapType)
 				if err != nil {
 					c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("failed to add annotation to app %#q in namespace %#q", app.Name, app.Namespace), "stack", fmt.Sprintf("%#v", err))
 					continue
 				}
 
 				c.logger.Debugf(ctx, "triggered %#q app update in namespace %#q", app.Name, app.Namespace)
+
+				c.event.Emit(ctx, currentApp, "AppUpdated", "triggered an update following to configmap %s/%s update", configMap.Namespace, configMap.Name)
```
Comment (tomahawk28) [inline service/watcher/appvalue/configmap.go:122]: I think I need a help from 🇬🇧 here.

**Comment by tomahawk28** [comment]
Not sure why appCatalogEntry test is failing but it seems not relevant to this PR.

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,10 @@ and this project's packages adheres to [Semantic Versioning](http://semver.org/s
 
 ## [Unreleased]
 
+### Added
+
+- Emit events when configMaps/secrets referenced in App CRs updated.
```
Comment (rossf7) [inline CHANGELOG.md:12]:
```suggestion
- Emit events when config maps and secrets referenced in App CRs are updated.
```

File: service/service.go [hunk 3]
```diff
@@ -122,9 +124,21 @@ func New(config Config) (*Service, error) {
 		}
 	}
 
+	var event recorder.Interface
+	{
+		c := recorder.Config{
+			K8sClient: config.K8sClient,
+
+			Component: fmt.Sprintf("%s-%s", project.Name(), project.Version()),
```
Comment (rossf7) [inline service/service.go:132]: Having the version too is really nice. ❤️

File: service/watcher/appvalue/configmap.go [hunk 4]
```diff
@@ -104,13 +105,21 @@ func (c *AppValueWatcher) watchConfigMap(ctx context.Context) {
 			for app := range storedIndex {
 				c.logger.Debugf(ctx, "triggering %#q app update in namespace %#q", app.Name, app.Namespace)
 
-				err := c.addAnnotation(ctx, app, cm.GetResourceVersion(), configMapType)
+				currentApp, err := c.k8sClient.G8sClient().ApplicationV1alpha1().Apps(app.Namespace).Get(ctx, app.Name, metav1.GetOptions{})
+				if err != nil {
+					c.logger.Errorf(ctx, err, "cannot fetch an app CR %s/%s", app.Namespace, app.Name)
```
Comment (rossf7) [inline service/watcher/appvalue/configmap.go:110]:
```suggestion
					c.logger.Errorf(ctx, err, "cannot fetch app CR %s/%s", app.Namespace, app.Name)
```

File: service/watcher/appvalue/secret.go [hunk 5]
```diff
@@ -100,13 +100,21 @@ func (c *AppValueWatcher) watchSecret(ctx context.Context) {
 			for app := range storedIndex {
 				c.logger.Debugf(ctx, "triggering %#q app update in namespace %#q", app.Name, app.Namespace)
 
-				err := c.addAnnotation(ctx, app, secret.GetResourceVersion(), secretType)
+				currentApp, err := c.k8sClient.G8sClient().ApplicationV1alpha1().Apps(app.Namespace).Get(ctx, app.Name, metav1.GetOptions{})
+				if err != nil {
+					c.logger.Errorf(ctx, err, "cannot fetch an app CR %s/%s", app.Namespace, app.Name)
```
Comment (rossf7) [inline service/watcher/appvalue/secret.go:105]:
```suggestion
					c.logger.Errorf(ctx, err, "cannot fetch app CR %s/%s", app.Namespace, app.Name)
```

File: service/watcher/appvalue/secret.go [hunk 6]
```diff
@@ -100,13 +100,21 @@ func (c *AppValueWatcher) watchSecret(ctx context.Context) {
...
+
+				err = c.addAnnotation(ctx, currentApp, secret.GetResourceVersion(), secretType)
 				if err != nil {
 					c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("failed to add annotation to app %#q in namespace %#q", app.Name, app.Namespace), "stack", fmt.Sprintf("%#v", err))
 					continue
 				}
 
 				c.logger.Debugf(ctx, "triggered %#q app update in namespace %#q", app.Name, app.Namespace)
+
+				c.event.Emit(ctx, currentApp, "AppUpdated", "triggered an update following to secret %s/%s update", secret.Namespace, secret.Name)
```
Comment (rossf7) [inline service/watcher/appvalue/secret.go:117]:
```suggestion
				c.event.Emit(ctx, currentApp, "AppUpdated", "change to secret %s/%s triggered an update", secret.Namespace, secret.Name)
```

File: service/watcher/appvalue/configmap.go (hunk 1 above)
Comment (rossf7) [inline service/watcher/appvalue/configmap.go:122]:
```suggestion
				c.event.Emit(ctx, currentApp, "AppUpdated", "change to configmap %s/%s triggered an update", configMap.Namespace, configMap.Name)
```

WDYT?

**Review by rossf7 — APPROVED** [review]
LGTM. Just a few minor things.

For the integration test it's really dumb. It's because of the latest label in the ACE entry we're checking.

Can you check? https://github.com/giantswarm/app-operator/pull/754

I did a quick fix to make it pass. I'll sort it out properly next week.

File: service/watcher/appvalue/configmap.go (hunk 1 above)
Comment (tomahawk28) [reply service/watcher/appvalue/configmap.go:122]: Much better 👍 thanks!

**Pull request merged**